src/ai/defaultai.cc

```diff
@@ -6137,15 +6138,16 @@ BuildingNecessity DefaultAI::check_building_necessity(BuildingObserver& bo,
 			return BuildingNecessity::kForbidden;
 		}
 		if (bo.is(BuildingAttribute::kShipyard)) {
-			if (bo.total_count() > 0 ||
+			if ((num_ports < 3 ? bo.total_count() > num_ports : bo.total_count() > num_ports / 2) ||
```

Comment:
```suggestion
			if (bo.total_count() > num_ports / 2 + 1) ||
```

Not only simpler, but also:

np | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
if | 0 | 1 | 2 | 1 | 2 | 2 | 3 | 3 | 4 | 4 | 5
n/2+1 | 1 | 1 | 2 | 2 | 3 | 3 | 4 | 4 | 5 | 5 | 6

Reply:
Sorry, this is post fact, so this may be better:

```suggestion
			if (bo.total_count() > (num_ports + 1) / 2) ||
```

np | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
if | 0 | 1 | 2 | 1 | 2 | 2 | 3 | 3 | 4 | 4 | 5
n/2+1 | 0 |1 | 1 | 2 | 2 | 3 | 3 | 4 | 4 | 5 | 5